### PR TITLE
Store Tracking Number

### PR DESCRIPTION
**Pull request by usudaysingh**

### CODE REVIEW

File: shipment.py [hunk 1]
```diff
@@ -509,3 +509,39 @@ def generate_label(self, shipment):
             )
 
         return getattr(shipment, method_name)()
+
+
+class ShipmentTracking(ModelSQL, ModelView):
+
```
Comment (priyankarani) [inline shipment.py]: remove blank line

File: shipment.py [hunk 2]
```diff
@@ -509,3 +509,39 @@ def generate_label(self, shipment):
...
+                states={
+                    'readonly': Eval('state') != 'draft',
+        },
+        domain=[
+            ('id', If(Eval('context', {}).contains('company'), '=', '!='),
+                Eval('context', {}).get('company', -1)),
+            ],
+        depends=['state'])
+
+    moves = fields.One2Many('stock.move', 'shipment', 'Moves',
```
Comment (priyankarani) [inline shipment.py]: move everything after '(' to next line

File: shipment.py [hunk 3]
```diff
@@ -509,3 +509,39 @@ def generate_label(self, shipment):
...
+        domain=[('company', '=', Eval('company'))], depends=['company'],
+        readonly=True)
```
Comment (priyankarani) [inline shipment.py]: move ')' to next line

File: shipment.py [hunk 4]
```diff
@@ -509,3 +509,39 @@ def generate_label(self, shipment):
...
+
+    origins = fields.Function(fields.Char('Origins'), 'get_origins')
```
Comment (priyankarani) [inline shipment.py]: field name is origin and it should be reference field

File: shipment.py [hunk 5]
```diff
@@ -509,3 +509,39 @@ def generate_label(self, shipment):
...
+        Save Tracking NUmber
+    """
+
+    __name__ = 'shipment.tracking'
+
+    _rec_name = 'tracking_number'
+
+    origin = fields.Reference(
+        'Origin', selection='get_origin', select=True,
+        states={
```
Comment (prakashpp) [inline shipment.py]: remove

File: shipment.py [hunk 6]
```diff
@@ -509,3 +509,39 @@ def generate_label(self, shipment):
...
+            'readonly': Eval('state') != 'draft',
+            },
+        depends=['state']
```
Comment (prakashpp) [inline shipment.py]: remove

File: shipment.py [hunk 7]
```diff
@@ -509,3 +509,35 @@ def generate_label(self, shipment):
             )
 
         return getattr(shipment, method_name)()
+
+
+class ShipmentTracking(ModelSQL, ModelView):
+    """
+        Save Tracking NUmber
+    """
+
```
Comment (prakashpp) [inline shipment.py]: remove blank line

File: shipment.py [hunk 8]
```diff
@@ -509,3 +509,35 @@ def generate_label(self, shipment):
             )
 
         return getattr(shipment, method_name)()
+
+
+class ShipmentTracking(ModelSQL, ModelView):
+    """
```
Comment (prakashpp) [inline shipment.py]: description in one line

File: shipment.py [hunk 9]
```diff
@@ -509,3 +509,35 @@ def generate_label(self, shipment):
...
+        Save Tracking NUmber
+    """
+
+    __name__ = 'shipment.tracking'
+
```
Comment (prakashpp) [inline shipment.py]: remove blank line

File: shipment.py [hunk 10]
```diff
@@ -509,3 +509,35 @@ def generate_label(self, shipment):
...
+    _rec_name = 'tracking_number'
+
+    origin = fields.Reference(
+        'Origin', selection='get_origin', select=True,
+    )
+
+    tracking_number = fields.Char("Tracking Number", required=True)
+
+    carrier = fields.Many2One('carrier', 'Carrier', required=True)
+
```
Comment (prakashpp) [inline shipment.py]: also add tracking url

Comment (usudaysingh) [inline shipment.py]: Not defined in task.Please define the properties too.

File: shipment.py [hunk 11]
```diff
@@ -509,3 +509,32 @@ def generate_label(self, shipment):
             )
 
         return getattr(shipment, method_name)()
+
+
+class ShipmentTracking(ModelSQL, ModelView):
+    """Save Tracking NUmber
```
Comment (tarunbhardwaj) [inline shipment.py]: s/U/u

File: shipment.py [hunk 12]
```diff
@@ -110,8 +110,17 @@ class ShipmentOut:
...
+    tracking_number = fields.Function(
+          fields.Many2One('shipment.tracking', 'Tracking Number'),
+          'get_tracking_number'
+    )
+
+
+    def get_tracking_number(self):
+        """
+        Return first tracking number
+        """
```
Comment (priyankarani) [inline shipment.py]: search for records in 'shipment.tracking' ( which would be list of active records )and return first shipment_tracking record here

File: shipment.py [hunk 13]
```diff
@@ -509,3 +518,34 @@ def generate_label(self, shipment):
...
+        'Return list of Model names for origin Reference'
+        return ['stock.shipment.out', 'stock.package']
+
+    @classmethod
+    def get_origin(cls):
+        Model = Pool().get('ir.model')
+        models = cls._get_origin()
+        models = Model.search([
+            ('model', 'in', models),
+            ])
```
Comment (priyankarani) [inline shipment.py]: deindent braces

**Comment by priyankarani** [comment]
Add tests

**Comment by usudaysingh** [comment]
for 3.4 https://github.com/fulfilio/trytond-shipping/pull/52

**Pull request closed without merging**